x-pack/solutions/security/plugins/security_solution/server/lib/siem_migrations/rules/index_migrator/per_space.ts

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type { ElasticsearchClient, Logger } from '@kbn/core/server';
+import type {
+  AggregationsAggregationContainer,
+  AggregationsMinAggregate,
+  AggregationsStringTermsAggregate,
+  AggregationsStringTermsBucket,
+} from '@elastic/elasticsearch/lib/api/types';
+import type { Adapters, StoredSiemMigration } from '../types';
+import { processResponseHits } from '../../common';
+
+const MAX_ES_SIZE = 10000;
+
+export class RuleMigrationSpaceIndexMigrator {
+  constructor(
+    private spaceId: string,
+    private esClient: ElasticsearchClient,
+    private logger: Logger,
+    private ruleMigrationIndexAdapters: Adapters
+  ) {}
+
+  private async getExistingMigrationFromRulesIndex() {
+    const index = this.ruleMigrationIndexAdapters.rules.getIndexName(this.spaceId);
+    const aggregations: Record<string, AggregationsAggregationContainer> = {
+      migrationIds: {
+        terms: { field: 'migration_id', order: { createdAt: 'asc' }, size: MAX_ES_SIZE },
+        aggregations: {
+          createdAt: { min: { field: '@timestamp' } },
+          createdBy: { terms: { field: 'created_by' } },
+        },
+      },
+    };
+    const result = await this.esClient
+      .search({ index, aggregations, _source: false })
+      .catch((error) => {
+        this.logger.error(`Error getting all rule migrations stats: ${error.message}`);
+        throw error;
+      });
+
+    const migrationsAgg = result.aggregations?.migrationIds as AggregationsStringTermsAggregate;
+    const buckets = (migrationsAgg?.buckets as AggregationsStringTermsBucket[]) ?? [];
+    return buckets.map((bucket) => ({
+      id: `${bucket.key}`,
+      created_at: (bucket.createdAt as AggregationsMinAggregate | undefined)
+        ?.value_as_string as string,
+      created_by: (
+        (bucket.createdBy as AggregationsStringTermsAggregate)
+          .buckets as AggregationsStringTermsBucket[]
+      )[0].key as string,
+    }));
+  }
+
+  private async getExistingMigrationFromMigrationsIndex() {
+    const index = this.ruleMigrationIndexAdapters.migrations.getIndexName(this.spaceId);
+    const result = await this.esClient.search<StoredSiemMigration>({
+      index,
+      query: {
+        match_all: {},
+      },
+      _source: true,
+    });
+    return processResponseHits(result);
+  }
```

Comment:
The only thing we care about here is the `id` string, I don't think we need to extract `processResponseHits` for this. We could use the logic we need:

```suggestion
  private async getExistingMigrationIdsFromMigrationsIndex(): Promise<string[]> {
    const index = this.ruleMigrationIndexAdapters.migrations.getIndexName(this.spaceId);
    const result = await this.esClient.search<StoredSiemMigration>({
      index,
      query: { match_all: {} },
      _source: true,
    });
    return result.hits.hits.map(({ _id }) => {
      assert(_id, 'document should have _id');
      return _id;
    });
  }
```

And adapt the code down below in the `migrateRuleMigrationIndex`

Reply:
Done here: https://github.com/elastic/kibana/pull/219597/commits/a3415aa25260ab3f2a66fc8da26ceaf616356895